lib/df-detection.c

```diff
@@ -0,0 +1,282 @@
+		if (current->addresses[current->pairs[i].first].stack &&
+		    current->addresses[current->pairs[i].second].stack) {
+			first_nr_entries = stack_depot_fetch(
+			    current->addresses[current->pairs[i].first].stack,
+			    &first_entries);
+			second_nr_entries = stack_depot_fetch(
+			    current->addresses[current->pairs[i].second].stack,
+			    &second_entries);
+			int first_index =
+			    filter_stack(first_entries, first_nr_entries);
+			int second_index =
+			    filter_stack(second_entries, second_nr_entries);
+			first_frame = first_entries[first_index];
+			second_frame = second_entries[second_index];
+			char first_buf[64], second_buf[64];
+			int first_len, second_len;
+			/*check if this bug happent in perf_copy_attr*/
+			first_len = scnprintf(first_buf, sizeof(first_buf),
+					      "%ps", (void *)first_frame);
+			second_len = scnprintf(second_buf, sizeof(second_buf),
+					       "%ps", (void *)second_frame);
+			if (strnstr(first_buf, "perf_copy_attr", first_len) &&
+			    strnstr(second_buf, "perf_copy_attr", second_len))
+				continue;
+			pr_err("BUG: multi-read in %ps  "
+			       "between %ps and %ps\n ",
+			       sys_call_table[current->syscall_num],
+			       first_frame, second_frame);
+			pr_err("==============================================="
+			       "===================\n");
+			pr_err("======= First Address Range Stack =======");
+			stack_trace_print(first_entries, first_nr_entries, 0);
+			pr_err("======= Second Address Range Stack =======");
+			stack_trace_print(second_entries, second_nr_entries, 0);
+		} else {
+			pr_err("BUG: Intersection Detected at syscall: %ps\n ",
+			       sys_call_table[current->syscall_num]);
+			pr_err("==============================================="
+			       "===================\n");
+		}
+		pr_err("syscall number %ld  System Call: %pSR\n",
+		       current->syscall_num,
+		       sys_call_table[current->syscall_num]);
+		pr_err(
+		    "First %px len %lu Caller %pSR \nSecond %px len "
+		    "%lu Caller %pSR \n",
+		    current->addresses[current->pairs[i].first].start_address,
+		    current->addresses[current->pairs[i].first].len,
+		    current->addresses[current->pairs[i].first].caller,
+		    current->addresses[current->pairs[i].second].start_address,
+		    current->addresses[current->pairs[i].second].len,
+		    current->addresses[current->pairs[i].second].caller);
+		pr_err("======================================================="
+		       "===========\n");
+		if (panic_on_warn) {
+			panic_on_warn = 0;
+			panic("panic_on_warn set. \n");
+		}
+	}
+}
+
+int filter_stack(const unsigned long stack_entries[], int num_entries)
+{
+	char buf[64];
+	int len, indx;
+	/*we are not interested in first 2 entries as they are always
+	 * add_address & df_save_stack*/
+	for (indx = 2; indx < num_entries; indx++) {
+		len = scnprintf(buf, sizeof(buf), "%ps",
+				(void *)stack_entries[indx]);
+
+		if (strnstr(buf, "copy_from_user", len) ||
```

Comment:
Dmitry says this filtering can be moved to pkg/report in syzkaller, that is probably also worth looking into.